### PR TITLE
*: add comparable varint encoding.

### PR DESCRIPTION
**Pull request by coocood**
It will be used to encode key with less bytes.

### CODE REVIEW

File: util/codec/number.go [hunk 1]
```diff
@@ -155,3 +156,125 @@ func DecodeUvarint(b []byte) ([]byte, uint64, error) {
...
+	negativeTagEnd   = 8        // Negative tag is (negativeTagEnd - length).
+	positiveTagStart = 0xff - 8 // Positive tag is (positiveTagStart + length).
+)
+
+// EncodeComparableVarint encodes a int64 to a mem-comparable bytes.
+func EncodeComparableVarint(b []byte, v int64) []byte {
+	if v < 0 {
+		// All negative value has a tag byte prefix (negativeTagEnd - length).
+		// Smaller negative value encodes to more bytes, has smaller tag.
+		if v >= 0xff {
```
Comment (hanfei1991) [inline util/codec/number.go]: let v = -v

**Comment by coocood** [comment]
@zimulala  @hanfei1991 PTAL

File: util/codec/number.go [hunk 2]
```diff
@@ -155,3 +156,125 @@ func DecodeUvarint(b []byte) ([]byte, uint64, error) {
...
+	} else {
+		length = int(first) - positiveTagStart
+	}
+	if len(b) < length {
+		return nil, 0, errors.Trace(errDecodeInsufficient)
+	}
+	for _, c := range b[:length] {
+		v = (v << 8) | uint64(c)
+	}
+	if first > positiveTagStart && v > math.MaxInt64 {
```
Comment (GregoryIan) [inline util/codec/number.go]: v is uint64

Comment (coocood) [inline util/codec/number.go]: @hanfei1991 
If that happens, the encoding is invalid, we can add extra check.

**Comment by hanfei1991** [comment]
LGTM

File: util/codec/number.go [hunk 3]
```diff
@@ -155,3 +156,127 @@ func DecodeUvarint(b []byte) ([]byte, uint64, error) {
...
+	}
+	length := int(first) - positiveTagStart
+	if len(b) < length {
+		return nil, 0, errors.Trace(errDecodeInsufficient)
+	}
+	var v uint64
+	for _, c := range b[:length] {
+		v = (v << 8) | uint64(c)
+	}
+	return b, v, nil
```
Comment (zimulala) [inline util/codec/number.go]: b ->  b[:length]

**Comment by zimulala** [comment]
LGTM

**Pull request merged**